File: app/src/main/java/me/pwcong/jpstart/manager/DBManager.java
package me.pwcong.jpstart.manager;

import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import me.pwcong.jpstart.App;
import me.pwcong.jpstart.R;
import me.pwcong.jpstart.comparator.JPItemComporator;
import me.pwcong.jpstart.conf.Constants;
import me.pwcong.jpstart.db.JPStartDatabase;
import me.pwcong.jpstart.mvp.bean.JPItem;
import me.pwcong.jpstart.utils.ResourceUtils;

/**
 * Created by Pwcong on 2016/9/25.
 */

public class DBManager {

    private static DBManager instance = null;

    private List<JPItem> query = null;
    private List<JPItem> qingYin = null;
    private List<JPItem> zhuoYin = null;
    private List<JPItem> aoYin = null;
    private List<JPItem> qingYinWithoutHeader = null;
    private List<JPItem> zhuoYinWithoutHeader = null;
    private List<JPItem> aoYinWithoutHeader = null;

    private DBManager() {
    }

    public static synchronized DBManager getInstance(){

        if(instance==null){
            instance=new DBManager();
        }

        return instance;
    }

    public void init(){

        getQingYinWithoutHeader();
        getZhuoYinWithoutHeader();
        getAoYinWithoutHeader();

    }


    public synchronized List<JPItem> query(){

        if(query==null){
            SQLiteDatabase db = new JPStartDatabase(App.getInstance()).getReadableDatabase();
            Cursor cursor = db.rawQuery("select * from " + JPStartDatabase.DB_TABLE_NAME, null);
            query = new ArrayList<>();
            JPItem item;
            while (cursor.moveToNext()){
                int id = cursor.getInt(cursor.getColumnIndex("id"));
                int row = cursor.getInt(cursor.getColumnIndex("row"));
                int column = cursor.getInt(cursor.getColumnIndex("column"));
                String hiragana = cursor.getString(cursor.getColumnIndex("hiragana"));
                String katakana = cursor.getString(cursor.getColumnIndex("katakana"));
                String rome = cursor.getString(cursor.getColumnIndex("rome"));
                int category = cursor.getInt(cursor.getColumnIndex("category"));
                int type = cursor.getInt(cursor.getColumnIndex("type"));
                boolean existed = cursor.getInt(cursor.getColumnIndex("existed"))==1;

                item=new JPItem(id,row,column,hiragana,katakana,rome,category,type,existed);

                query.add(item);
            }
            cursor.close();
            db.close();
        }

        return query;
    }

    public synchronized List<JPItem> getQingYin() {

        if (qingYin==null){

            qingYin=new ArrayList<>();
            List<JPItem> query = query();
            for(JPItem item:query){

                if(item.getCategory()==Constants.CATEGORY_QINGYIN){
                    qingYin.add(item);
                }

            }

            Collections.sort(qingYin,new JPItemComporator());


        }

        return qingYin;
    }

    public List<JPItem> getZhuoYin() {

        if (zhuoYin==null){

            zhuoYin=new ArrayList<>();
            List<JPItem> query = query();
            for(JPItem item:query){

                if(item.getCategory()==Constants.CATEGORY_ZHUOYIN){
                    zhuoYin.add(item);
                }

            }

            Collections.sort(zhuoYin,new JPItemComporator());



        }


        return zhuoYin;
    }

    public List<JPItem> getAoYin() {

        if (aoYin==null){

            aoYin=new ArrayList<>();
            List<JPItem> query = query();
            for(JPItem item:query){

                if(item.getCategory()==Constants.CATEGORY_AOYIN){
                    aoYin.add(item);
                }

            }

            Collections.sort(aoYin,new JPItemComporator());


        }

        return aoYin;
    }

    public List<JPItem> getQingYinWithoutHeader() {

        if (qingYinWithoutHeader==null){

            qingYinWithoutHeader=new ArrayList<>();
            List<JPItem> query = getQingYin();
            for(JPItem item:query){

                if(item.getRow()!=0&&item.getColumn()!=0){
                    qingYinWithoutHeader.add(item);
                }

            }
        }



        return qingYinWithoutHeader;
    }

    public List<JPItem> getZhuoYinWithoutHeader() {

        if (zhuoYinWithoutHeader==null){

            zhuoYinWithoutHeader=new ArrayList<>();
            List<JPItem> query = getZhuoYin();
            for(JPItem item:query){

                if(item.getRow()!=0&&item.getColumn()!=0){
                    zhuoYinWithoutHeader.add(item);
                }

            }
        }

        return zhuoYinWithoutHeader;
    }

    public List<JPItem> getAoYinWithoutHeader() {

        if (aoYinWithoutHeader==null){

            aoYinWithoutHeader=new ArrayList<>();
            List<JPItem> query = getAoYin();
            for(JPItem item:query){

                if(item.getRow()!=0&&item.getColumn()!=0){
                    aoYinWithoutHeader.add(item);
                }

            }
        }

        return aoYinWithoutHeader;
    }

    public static void addHeaderString(List<JPItem> list, int row, int column){

        for(int i=1;i<column;i++){

            JPItem item = list.get(i);
            String hiragana = item.getHiragana();
            String katakana = item.getKatakana();
            item.setHiragana(hiragana+ResourceUtils.getString(App.getInstance(),R.string.column));
            item.setKatakana(katakana+ResourceUtils.getString(App.getInstance(),R.string.column));

        }

        for(int i = 1;i<row;i++){

            JPItem item = list.get(i * column);
            String hiragana = item.getHiragana();
            String katakana = item.getKatakana();
            item.setHiragana(hiragana+ResourceUtils.getString(App.getInstance(),R.string.row));
            item.setKatakana(katakana+ResourceUtils.getString(App.getInstance(),R.string.row));

        }

    }



}

File: app/src/main/java/com/github/pwcong/jpstart/manager/DBManager.kt
package com.github.pwcong.jpstart.manager

import java.util.Collections

import android.database.sqlite.SQLiteDatabase

import com.github.pwcong.jpstart.App
import com.github.pwcong.jpstart.R
import com.github.pwcong.jpstart.comparator.JPItemComparator
import com.github.pwcong.jpstart.constants.Constants
import com.github.pwcong.jpstart.db.JPStartDatabase
import com.github.pwcong.jpstart.mvp.bean.JPItem
import com.github.pwcong.jpstart.utils.ResourceUtils

class DBManager private constructor() {
    private var query: MutableList<JPItem>? = null
    private var qingYin: MutableList<JPItem>? = null
    private var zhuoYin: MutableList<JPItem>? = null
    private var aoYin: MutableList<JPItem>? = null
    private var qingYinWithoutHeader: MutableList<JPItem>? = null
    private var zhuoYinWithoutHeader: MutableList<JPItem>? = null
    private var aoYinWithoutHeader: MutableList<JPItem>? = null

    fun init() {
        getQingYinWithoutHeader()
        getZhuoYinWithoutHeader()
        getAoYinWithoutHeader()
    }

    @Synchronized
    fun query(): List<JPItem> {
        if (query == null) {
            val db: SQLiteDatabase = JPStartDatabase(App.getInstance()).getReadableDatabase()
            val cursor = db.rawQuery("select * from " + JPStartDatabase.DB_TABLE_NAME, null)
            query = mutableListOf<JPItem>()
            var item: JPItem
            while (cursor.moveToNext()) {
                val idIndex = cursor.getColumnIndex("id")
                val id = cursor.getInt(idIndex)

                val rowIndex = cursor.getColumnIndex("row")
                val row = cursor.getInt(rowIndex)

                val columnIndex = cursor.getColumnIndex("column")
                val column = cursor.getInt(columnIndex)

                val hiraganaIndex = cursor.getColumnIndex("hiragana")
                val hiragana = cursor.getString(hiraganaIndex)

                val katakanaIndex = cursor.getColumnIndex("katakana")
                val katakana = cursor.getString(katakanaIndex)

                val romeIndex = cursor.getColumnIndex("rome")
                val rome = cursor.getString(romeIndex)

                val categoryIndex = cursor.getColumnIndex("category")
                val category = cursor.getInt(categoryIndex)

                val typeIndex = cursor.getColumnIndex("type")
                val type = cursor.getInt(typeIndex)

                val existedIndex = cursor.getColumnIndex("existed")
                val existed = cursor.getInt(existedIndex) == 1

                item = JPItem(id, row, column, hiragana, katakana, rome, category, type, existed)

                query?.add(item)
            }
            cursor.close()
            db.close()
        }

        return query!!
    }

    @Synchronized
    fun getQingYin(): List<JPItem> {
        if (qingYin == null) {
            qingYin = mutableListOf()
            val query = query()
            for (item in query) {
                if (item.category == Constants.CATEGORY_QINGYIN) {
                    qingYin!!.add(item)
                }
            }

            Collections.sort<JPItem>(qingYin!!, JPItemComparator())
        }

        return qingYin!!
    }

    @Synchronized
    fun getZhuoYin(): List<JPItem> {
        if (zhuoYin == null) {
            zhuoYin = mutableListOf()
            val query = query()
            for (item in query) {
                if (item.category == Constants.CATEGORY_ZHUOYIN) {
                    zhuoYin!!.add(item)
                }
            }

            Collections.sort<JPItem>(zhuoYin!!, JPItemComparator())
        }

        return zhuoYin!!
    }

    @Synchronized
    fun getAoYin(): List<JPItem> {
        if (aoYin == null) {
            aoYin = mutableListOf()
            val query = query()
            for (item in query) {
                if (item.category == Constants.CATEGORY_AOYIN) {
                    aoYin!!.add(item)
                }
            }

            Collections.sort<JPItem>(aoYin!!, JPItemComparator())
        }

        return aoYin!!
    }

    fun getQingYinWithoutHeader(): List<JPItem> {
        if (qingYinWithoutHeader == null) {
            qingYinWithoutHeader = mutableListOf()
            val query = getQingYin()
            for (item in query) {
                if (item.row != 0 && item.column != 0 && item.isExisted) {
                    qingYinWithoutHeader!!.add(item)
                }
            }
        }

        return qingYinWithoutHeader!!
    }

    fun getZhuoYinWithoutHeader(): List<JPItem> {
        if (zhuoYinWithoutHeader == null) {
            zhuoYinWithoutHeader = mutableListOf()
            val query = getZhuoYin()
            for (item in query) {
                if (item.row != 0 && item.column != 0 && item.isExisted) {
                    zhuoYinWithoutHeader!!.add(item)
                }
            }
        }

        return zhuoYinWithoutHeader!!
    }

    fun getAoYinWithoutHeader(): List<JPItem> {
        if (aoYinWithoutHeader == null) {
            aoYinWithoutHeader = mutableListOf()
            val query = getAoYin()
            for (item in query) {
                if (item.row != 0 && item.column != 0 && item.isExisted) {
                    aoYinWithoutHeader!!.add(item)
                }
            }
        }

        return aoYinWithoutHeader!!
    }

    companion object {
        private var instance: DBManager? = null

        @Synchronized
        fun getInstance(): DBManager {
            if (instance == null) {
                instance = DBManager()
            }

            return instance!!
        }

        fun addHeaderString(list: List<JPItem>, row: Int, column: Int) {
            for (i in 1 until column) {
                val item = list[i]
                val hiragana = item.hiragana
                val katakana = item.katakana
                item.hiragana =
                    hiragana + ResourceUtils.getString(App.getInstance(), R.string.column)
                item.katakana =
                    katakana + ResourceUtils.getString(App.getInstance(), R.string.column)
            }

            for (i in 1 until row) {
                val item = list[i * column]
                val hiragana = item.hiragana
                val katakana = item.katakana
                item.hiragana = hiragana + ResourceUtils.getString(App.getInstance(), R.string.row)
                item.katakana = katakana + ResourceUtils.getString(App.getInstance(), R.string.row)
            }
        }
    }
}